src/compiler/checker.ts

```diff
@@ -24066,7 +24078,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         const templateType = getTemplateTypeFromMappedType(target);
         const inference = createInferenceInfo(typeParameter);
         inferTypes([inference], sourceType, templateType);
-        return getTypeFromInference(inference) || unknownType;
+        // Skip subtype reducing the inference result (this can eagerly resolve structure, which we'd rather avoid while doing this inference)
+        return getTypeFromInference(inference, UnionReduction.Literal) || unknownType;

```

Comment:
By only doing `Literal` reductions on this inference result, we save some time, and prevent pulling on the structure of a recursive reverse mapped type while it's still being constructed. What we lose is, conceivably, reducing a reverse mapped type property that resolves to `{x} | {x,y}` to `{x,y}` (Or, perhaps more relevant, `Function | RecursiveReverseMappingThatResolvesToASpecificFunction` becoming `Function`, which is closer to what `mongoose` has); but I'd argue that _most_ reverse mapped types only involve a single type, maybe union'd with `undefined` anyway.

The alternative is just issuing an error on the recursion, and that code is in-place, but that error feels bad because it only appears on instantiation and usage of the reverse mapped type, and not when it's made, even though the error appears at its' declaration site.

(note: as an aside, the command-line compiler does _not_ like this and won't display the error, since it's generated for a file that already had direct diagnostics cached, resulting in a very silent, sneaky `any` type popping out of inference without an error reported. I suspect any deferred structure resolution that has a circularity can trigger this poor error reporting behavior. I can dodge this by using `currentNode` as the error node for the circularity error instead of the symbol's declaration site, but then you loose where the circularity occurs, so it probably at least needs a related span... better would be fixing up the program to actually pick up diagnostics added in files which it already collected diagnostics in once.)